### DISCUSSION TITLE
Option to left align table captions

### DISCUSSION THREAD

**Discussion by arnaudgallou** [quarto-dev/quarto-cli#1042] [Q&A] [answered]
Is there a `yaml` option to left align table captions? By default, the captions of my tables are centered. This might seem trivial but I couldn't find the information in the documentation.

**Comment by jjallaire** [comment, marked as answer]
Currently you can do this with CSS. For example:

````markdown
```{=html}
<style type="text/css">
caption, .table-caption {
  text-align: left;
}
</style>
```
````

Alternatively you can place the CSS inside a styles.css file that you include with e.g. `css: styles.css`

**Reply by martinschaffhauser** [reply]
Can anyone confirm that adding

```
caption, .table-caption {
  text-align: left;
}
```

in the `styles.css` for rendering to html works with `quarto 1.4.555`? 

For me that does not affect table caption alignment - it stays centered.

On the other hand, adding
```
figcaption {
  margin: auto;
  text-align: right;
}
```
to `styles.css` does in fact change the figure caption alignment as desired.

Does anyone have a working solution for left aligning table captions in html documents?

**Reply by mcanouil** [reply]
@martinschaffhauser We unfortunately can't because we have no code to work with.

Could you share a **small self-contained "working" (reproducible)** example to work with, _i.e._, a complete Quarto document or a Git repository? The goal is to make it as easy as possible for us to recreate your problem so that we can fix it: please help us help you! Thanks.

---

You can share a **self-contained "working" (reproducible)** Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).
See <https://quarto.org/bug-reports.html#small-is-beautiful-aim-for-a-single-document-with-10-lines>.

If you have multiple files (and if it is **absolutely required** to have multiple files), please share as a Git repository.

<table>
<tr><th>R</th><th>Python</th></tr>
<tr><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{r}
x <- c(1, 2, 3, 4, 5)
y <- c(1, 4, 9, 16, 25)

plot(x, y)
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{python}
import matplotlib.pyplot as plt

x = [1, 2, 3, 4, 5]
y = [1, 4, 9, 16, 25]

plt.plot(x, y)
plt.show()
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td></tr>
</table>

Additionally and if not already given, please share the output of `quarto check` within a code blocks (*i.e.*, using three backticks ` ```txt `), see <https://quarto.org/bug-reports.html#check>.

**Reply by martinschaffhauser** [reply]
Of course!

`reprex.qmd`:
````qmd
---
title: "reprex"
format: 
  html:
    page-layout: full
    toc: true
    toc-depth: 4
    number-sections: true
css: "stylereprex.css"
html-table-processing: none # https://quarto.org/docs/prerelease/1.4/ast.html#finer-control-over-table-processing
editor_options: 
  chunk_output_type: console
---

```{r libs}
#| echo: false
#| message: false
#| warning: false

library(kableExtra)
library(dplyr)
```

```{r data-reprex}
#| echo: false
#| message: false
#| warning: false

set.seed(20240828)

n <- 100
group <- c(rep("A", 53), rep("B", n - 53))
age <- rnorm(n, mean = 60, sd = 20)
outcome <- ifelse(group == "A", rnorm(53, mean = 100, sd = 15), rnorm(n - 53, mean = 120, sd = 15))
df <- data.frame(Group = group, Age = age, Outcome = outcome)
```

```{r table-reprex}
#| label: tbl-reprex
#| tbl-cap: "**Table reprex**. The caption should be left aligned."
#| echo: false
#| warning: false
#| message: false

df %>%
  group_by(Group) %>%
  summarise(`Mean Outcome` = mean(Outcome)) %>%
  mutate(across(where(is.numeric), ~ round(., 2))) %>%
  kable(.,
        row.names = F,
        escape = F) %>% 
        kable_styling(full_width = F, position = "left")
```

See table caption of @tbl-reprex. The caption should be left aligned.
````

`stylereprex.css`:
```
caption, .table-caption {
  text-align: left;
}

/* works for figure caption alignment as intended */
figcaption {
  margin: auto;
  text-align: right;
}
```

`quarto check`:
```txt
Quarto 1.4.555
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.11: OK
      Dart Sass version 1.69.5: OK
      Deno version 1.37.2: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.4.555
      Path: /usr/lib/rstudio-server/bin/quarto/bin

[✓] Checking tools....................OK
      TinyTeX: (not installed)
      Chromium: (not installed)

[✓] Checking LaTeX....................OK
      Using: Installation From Path
      Path: /usr/local/texlive/bin/linux
      Version: 2024

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.10.12
      Path: /usr/bin/python3
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with python3 -m pip install jupyter

[✓] Checking R installation...........OK
      Version: 4.4.1
      Path: /usr/local/lib/R
      LibPaths:
        - /usr/local/lib/R/site-library
        - /usr/local/lib/R/library
      knitr: 1.47
      rmarkdown: 2.27

[✓] Checking Knitr engine render......OK
```

Thanks for your input!

**Reply by martinschaffhauser** [reply]
So i just figured out a working `css` snippet:

```css
figure.quarto-float-tbl figcaption.quarto-float-caption-top {
  text-align: left;
}
```

I hope this helps anyone!

**Reply by mcanouil** [reply]
Thanks.

FYI, you are setting twice the code cell labels in your code.
The following is wrong (even in RMarkdown):
````qmd
```{r table-reprex}
#| label: tbl-reprex
````

You should not mix R inline code cell syntax and YAML syntax.
````qmd
```{r table-reprex}
````

The above is strictly equivalent to
````qmd
```{r}
#| label: table-reprex
````

Quarto recommends using only the YAML syntax, *i.e.*, it means you never have something next to the language within the brackets ` ```{LANGUAGE} `.

**Reply by martinschaffhauser** [reply]
Thank you for the clarification Mickaël - I am transitioning to Quarto from Rmarkdown. The `#|` syntax is completely new to me. So far it's quite the journey to change some of my existing hacky Rmarkdown reports so they render as intended.

**Reply by mcanouil** [reply]
`#|` is the recommended and most used syntax in Quarto, but `knitr`/`rmarkdown` supports it natively since few years now.

**Comment by mvanaman** [comment]
This is a great CSS solution. But given how common a top-left caption is, personally I would reeeeaallly like to see a "topleft" option for the `tbl-cap-location:` option

**Reply by schwa021** [reply]
It's strange that, even though it appears as "Table caption" style in word, it does not respect the "Table caption" paragraph style (e.g., left justified)

**Comment by eteitelbaum** [comment]
The proposed solution here would only work for .html documents, right? 

Also, what would be the equivalent code for figures or if you wanted to have left-justified captions for both figures and tables?

**Reply by mcanouil** [reply]
To left-align the caption element in LaTeX, you can use the `caption` package (used by Quarto in the template) and set the `justification` option to `raggedright`. Here's an example code snippet:

````qmd
---
title: "quarto-table"
format: pdf
keep-tex: true
include-before-body:
  - text: |
      \captionsetup{justification=raggedright,singlelinecheck=false}
---

| Col1 | Col2 | Col3 |
|------|------|------|
| A    | B    | C    |
| E    | F    | G    |
| A    | G    | G    |

: My Caption {#tbl-letters}

See @tbl-letters.
````

In this example, the `caption` package is loaded in the header by Quarto (see `keep-tex: true`), and the `justification` option is set to `raggedright` before the body after the `caption` package is loaded. This option causes the caption text to be left-aligned.

The `singlelinecheck` option is set to `false` to ensure that the caption is left-aligned even if it is a single line. If this option is set to `true`, the caption will be centred if it is a single line.

You can adjust the `justification` option to `centering` or `raggedleft` to centre or right-align the caption, respectively.

**Reply by eteitelbaum** [reply]
Thank you! The problem I am having is that Quarto aligns the caption with the margin rather than the edge of an image. So I have tried:

```
 \captionsetup{justification = raggedright, singlelinecheck = false}
```

and:


```
 \captionsetup{justification = justified, singlelinecheck = false}
```

In both cases, here is what I see: 

![big_image](https://github.com/quarto-dev/quarto-cli/assets/62634082/5feb54d5-320d-4489-8ef8-a1c995010278)

The problem is even worse when you reduce the size of the image: 

![reduced_size_image](https://github.com/quarto-dev/quarto-cli/assets/62634082/89e8bdcb-9e0b-425d-9744-256269144e58)

Ideally, you would want the caption to align with the edge of the image rather than the margin.

**Reply by eteitelbaum** [reply]
Actually, I am looking at some journals in my field now and and it seems like most of them align the captions with the margins rather than the edge of the figure. When I was looking at the Turabian manual yesterday I was seeing captions aligned with figures, but perhaps for most academic purposes the margin-based alignment is fine.

**Reply by mcanouil** [reply]
> Actually, I am looking at some journals in my field now and and it seems like most of them align the captions with the margins rather than the edge of the figure.

I don't believe it's easy to align to the edge in LaTeX (or any editing tools) automatically.

PS: do note that there are way better places to ask LaTeX questions than here, _i.e., I found the answer I gave on the Internet (not off the top of my head, even it was there somewhere initially).
This being said, here it is useful to complete the HTML answer with the LaTeX answer.

**Reply by eteitelbaum** [reply]
Yes I had also found the answer with ChatGPT by the time you had answered it @mcanouil. You are saying there is a rule or norm against asking LaTeX questions on the Quarto CLI discussion board?

**Reply by mcanouil** [reply]
What I am saying is: as far as I know, there is no "expert" LaTeX user here, so don't expect too much to get answer especially for "advanced" LaTeX questions.

Due to the multiple formats allowed by Quarto/Pandoc which all have their own specificity and syntax, dedicated channel on StackOverflow for example will likely be more helpful.

This repository is really about Quarto itself where there is all the expertise we can provide (note that we really try to answer all questions anyway the best we can).

**Reply by eteitelbaum** [reply]
I see what you mean. But do you think if I had taken this question to a LaTeX discussion board (e.g. how do I adjust the alignment of captions in Quarto) I would have gotten a good answer?

**Reply by mcanouil** [reply]
> This being said, here it is useful to complete the HTML answer with the LaTeX answer.

Quoting myself

**Comment by dmi3kno** [comment]
My LaTeX class aligns figure captions left, but does not affect the table caption alignment. I wonder if there's a way to follow the alignment for table captions as well... I am afraid that if you guys make `topleft` or `bottomleft` then it would not honor the margins or something like that. Speaking of that, I think caption location should be left as is and there there should another option `tbl-cap-align` with left, right, center as options. We will deal with margins manually (somehow).

**Comment by marioem** [comment]
Suggested workaround doesn't work for me with flextable.

This one works however (html output):

```html
<style type="text/css">
  figure.quarto-float-tbl figcaption.quarto-float-caption-top {
    margin-top: .5rem;
    margin-bottom: .25rem;
    text-align:left
  }
</style>
```
